index.bs

```diff
@@ -261,16 +282,33 @@ dictionary MediaStreamTrackGeneratorInit {
     with |init|.{{MediaStreamTrackGeneratorInit/kind}}.
 4. Return |g|.
 
-### Attributes ### {#attributes-generator}
+### Attributes ### {#generator-attributes}
 <dl>
 <dt><dfn attribute for=MediaStreamTrackGenerator>writable</dfn></dt>
-<dd>Allows writing media frames the {{MediaStreamTrackGenerator}}, which is
-itself a {{MediaStreamTrack}}. If the {{MediaStreamTrack/kind}} attribute is `"audio"`,
-the stream will accept {{AudioData}} objects and fail with any other type. If
-{{MediaStreamTrack/kind}} is `"video"`, the stream will accept {{VideoFrame}} objects
-and fail with any other type. When a frame is written to {{MediaStreamTrackGenerator/writable}},
-the frame's `close()` method is automatically invoked, so that its internal resources are no longer
-accessible from JavaScript.
+<dd>Allows writing media frames to the {{MediaStreamTrackGenerator}}, which is
+itself a {{MediaStreamTrack}}. When this attribute is access for the first time,
+it MUST be initialized with the following steps:
+1. Initialize [=this=].{{MediaStreamTrackGenerator/writable}} to be a new {{WritableStream}}.
+2. <a dfn for="WritableStream">Set up</a> [=this=].{{MediaStreamTrackGenerator/writable}}, with its [=WritableStream/set up/writeAlgorithm=] set to [=writeFrame=] with [=this=] as parameter, [=WritableStream/set up/closeAlgorithm=] set to [=closeWritable=] with [=this=] as parameter, and [=WritableStream/set up/abortAlgorithm=] set to [=closeWritable=] with [=this=] as parameter.
+
+The <dfn>writeFrame</dfn> algorithm is given a |generator| and a |frame| as input. It is defined by running the following steps:
+1. If |generator|.{{MediaStreamTrack/kind}} equals `video` and |frame| is not a {{VideoFrame}} object, return [=a promise rejected with=] a {{TypeError}}.
+2. If |generator|.{{MediaStreamTrack/kind}} equals `audio` and |frame| is not an {{AudioData}} object, return [=a promise rejected with=] a {{TypeError}}.
+3. Send the media data backing |frame| to all live tracks connected to |generator|, possibly including |generator| itself.
+4. Invoke the `close` method of |frame|.
+5. Return  [=a promise resolved with=] undefined.
+
+The <dfn>closeWritable</dfn> algorithm is given a |generator| as input.
+It is defined by running the following steps.
+1. {{StopSource|Stop}} the generator as a MediaStreamTrack source.
+
+<p class="note">
+A consecuence of this step is that all tracks connected to
+|generator|.{{MediaStreamTrackGenerator/writable}} will be ended.
+</p>
+2. Return [=a promise resolved with=] undefined.
+
+
```

Comment:
IMO it would be good to define algorithms for operations on MSTG-as-source. I know that "stop" is one such operation, but perhaps there are others?

Reply:
This would actually be a bug for mediacapture-main. This spec doesn't specify any source behavior different from any other source, so source operations should be defined there.
This spec only specifies when some of those operations need to occur (e.g., when closeAlgorithm is invoked).

Reply:
IMO, the difference between this source and other sources is that in this case the source is fully specified.

I looked through mediacapture-main, and the only other thing I noticed was muted state.

Note that you recently added https://w3c.github.io/mediacapture-transform/#generator-stop that repeats the stop algorithm.